### DISCUSSION TITLE
Observing multiple elements

### DISCUSSION THREAD

**Discussion by favourwright** [vueuse/vueuse#1922] [Q&A] [answered]
With regular intersection observer, I can observe multiple elements and run a callback when they're in the viewport but I can't seem to figure out how to go about it using the one provided in vueuse

anyone has an idea how?

**Comment by darrylnoakes** [comment, marked as answer]
The current implementation accepts a single target and observes just that target, so it is not currently possible.

However, the target parameter could be changed to accept an array of targets as well, and then it can observe all of them.
If someone wants to make a PR...

**Reply by lumdzeehol** [reply]
With your suggestion I have changed target so that it could pass both single ref or ref array. code looks like below:
```
        const observer = new IntersectionObserver(
          callback,
          {
            root,
            rootMargin,
            threshold,
          },
        )
        if (Array.isArray(el)) {
          el.forEach((elm) => {
            if (!elm)
              return
            observer.observe(elm)
          })
        }
        else {
          observer.observe(el)
        }
```

But I just found that if i passed a ref array as target into `useIntersectionObserver`, it's a little hard to manage the relation between IntersectionObserver callback entries  and target refs . the intersectionObserver observed those doms, but the callback entries just contain the entry that is intersecting the root.

Maybe I should handle it with a `Map<target, index>` inside this hook?

Looking forward to getting advice from you, thx

**Reply by darrylnoakes** [reply]
Using `IntersectionObserver` normally (without Vue), this sort of management would be handled using ids, classes, data attributes, etc.
There are many different paradigms and methods to do it.
This composable supposed to be a _wrapper_ for `IntersectionObserver`: that management should probably left to the user.

You could make a new "higher-level" composable that works a bit differently and provides additional features.
E.g.
Its return could include an array ref, with each item being an `IntersectionObserverEntry` (undefined if no data available) for the ref at the same index of the array passed as the target. Then you can watch that array for changes.
Or an object, keyed by ref somehow.
There are many ways, which is probably why nothing is currently provided.

**Reply by shtse8** [reply]
I would like to `useElementBounding` on ref of array of elements. is it possible now?

**Reply by darrylnoakes** [reply]
The best way for that would probably be to map the array through `useElementBounding`.
Something like:

```ts
// With computed + Array.map:
const rects = computed(() =>
  elements.value.map((el) => useElementBounding(el))
);

// Or with VueUse's `useArrayMap`:
const rects = useArrayMap(elements, (el) => useElementBounding(el));
```

**Reply by shtse8** [reply]
> The best way for that would probably be to map the array through `useElementBounding`. Something like:
> 
> ```ts
> // With computed + Array.map:
> const rects = computed(() =>
>   elements.value.map((el) => useElementBounding(el))
> );
> 
> // Or with VueUse's `useArrayMap`:
> const rects = useArrayMap(elements, (el) => useElementBounding(el));
> ```

it doesn't work because ref of array of elements may change after mounted. so `onMounted` inside the `useElementBounding` will be called after `setup` which is out of scope.

I re-implemented `useElementBounding` to support `ref of array of elements`.
```ts
import { Ref } from '@vue/reactivity'
import { MaybeComputedElementRef, tryOnScopeDispose, unrefElement } from '@vueuse/core'
import { MaybeRef, resolveUnref } from '@vueuse/shared'

interface Rect {
  height: number
  bottom: number
  left: number
  right: number
  top: number
  width: number
  x: number
  y: number
}

export function useElementsBounding(targets: MaybeRef<MaybeComputedElementRef[]>) {
  const map = computed(() => resolveUnref(targets).reduce((map, x) => {
    const el = unrefElement(x)
    if (el)
      map.set(el, ref<Rect>({
        height: 0,
        bottom: 0,
        left: 0,
        right: 0,
        top: 0,
        width: 0,
        x: 0,
        y: 0
      }))
    return map
  }, new Map<Element, Ref<Rect>>()))

  const boundings = computed(() => resolveUnref(targets)
    .map(x => unrefElement(x))
    .filter(x => x && map.value.has(x))
    .map(x => map.value.get(x!)!))

  function update(target: Element) {
    const refRect = map.value.get(target)
    if (refRect) {
      const rect = target.getBoundingClientRect()
      refRect.value.height = rect.height
      refRect.value.bottom = rect.bottom
      refRect.value.left = rect.left
      refRect.value.right = rect.right
      refRect.value.top = rect.top
      refRect.value.width = rect.x
      refRect.value.x = rect.x
      refRect.value.y = rect.y
    }
  }

  function updateAll() {
    for (const target of map.value.keys()) {
      update(target)
    }
  }

  tryOnMounted(() => {
    const observer = new ResizeObserver(entries => {
      for (const entry of entries) {
        update(entry.target)
      }
    })
    watch(() => resolveUnref(targets), (current, previous) => {
      if (previous) {
        for (const target of previous) {
          const el = unrefElement(target)
          if (el)
            observer.unobserve(el)
        }
      }
      for (const target of current) {
        const el = unrefElement(target)
        if (el)
          observer.observe(el)
      }
    }, {
      immediate: true
    })
    tryOnScopeDispose(() => observer.disconnect())
  })
  useEventListener('resize', updateAll, { passive: true })
  tryOnMounted(() => updateAll())
  return boundings
}

```
